### PR TITLE
feature: initialize the parent and attach the grandparent in attach_scope

### PR DESCRIPTION
**Pull request by adrianthedev**
# Description

Fixes https://github.com/avo-hq/avo/issues/904

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the [documentation](https://github.com/avo-hq/docs)
- [ ] I have added tests that prove my fix is effective or that my feature works

## Manual review steps

Description...

1. Step 1
1. Step 2

Manual reviewer: please leave a comment with output from the test if that's the case.

### CODE REVIEW

File: app/controllers/avo/search_controller.rb [hunk 1]
```diff
@@ -72,6 +61,31 @@ def search_resource(resource)
...
+
+      # Try to fetch the parent.
+      if params[:via_reflection_id].present?
+        parent = params[:via_reflection_class].safe_constantize.find params[:via_reflection_id]
+      end
+
+      # Try to fetch the grandparent for the new views where the parent is nil.
+      if params[:via_parent_resource_id].present? && params[:via_parent_resource_class].present?
+        grandparent = params[:via_parent_resource_class].safe_constantize.find params[:via_parent_resource_id]
+      end
```
Comment (davidlormor) [inline app/controllers/avo/search_controller.rb:83]: Hmmm...I wonder if it would be feasible to use `find_or_initialize_by` to pass in the association on the parent vs. introducing another variable? Would look something like:

```ruby
grandparent = params[:via_parent_resource_class].safe_constantize.find params[:via_parent_resource_id]

parent = params[:via_reflection_class].safe_constantize.find_or_initialize_by(
  "id" => params[:via_reflection_id],
  params[:via_parent_resource_relation] => grandparent
)
```

I'm not familiar enough with Avo's internals to know where the grandparent relation might come from, so not really sure on the variable naming...

Another thought might be to use [the `GlobalID` feature](https://github.com/rails/globalid/blob/main/README.md) introduced to Rails alongside ActiveJob to serialize the parent into a URI perhaps?

**Review by davidlormor — COMMENTED** [review]
@adrianthedev this appears to introduce a breaking change when `searchable: false`. I get the following error: `Avo::Hosts::AssociationScopeHost: option 'grandparent' is required`. What's also strange is I don't even see the `attach_scope` being applied when `searchable: true` is applied...if I try to `binding.pry` in the attach scope I'm never receiving a breakpoint to inspect what's going on? Really weird, since I'm pretty sure that I was able to do it yesterday...

Also, I agree the `grandparent` naming is a bit "awkward"...see my thoughts on how to possibly avoid introducing a new variable by utilizing `find_or_initialize_by`.

**Comment by davidlormor** [comment]
Also should've mentioned - thanks for looking into this! Keep up the amazing work!

File: app/controllers/avo/search_controller.rb (hunk 1 above)
Comment (adrianthedev) [reply app/controllers/avo/search_controller.rb]: Hey @davidlormor.

TL;DR: Try `2.6.1.pre.2` and use `parent.industry` in the attach scope. You'll have access to the `Industry` record.

Long version👇

In the scenario where you'd edit the `AdminAccount`, and try to change the `belongs_to` of the `person`, the parent is a record of `AdminAccount` class. So, in the same manner, when creating an `AdminAccount`, when setting the `person`, the `parent` will still have to be an `AdminAccount` instance, but there isn't one right now because you're just creating it.

I followed your suggestion and initialized the parent model `AdminAccount` with the grandparent attached. So in the attach scope you can use `parent.industry`. `parent` will be an `AdminAccount` without an ID, but with the `industry_id` attribute populated so you have all the info needed to scope the query.

I fixed the `searchable: false` bug too. Thanks for the shoutout.

**Comment by davidlormor** [comment]
@adrianthedev `2.6.1.pre.2` appears to work as expected 🎉

Thanks for the quick fixes - it's going to clean up both our codebase and admin user experience significantly!

**Comment by adrianthedev** [comment]
Coolio! I'm glad it's fixed. 
I'll merge and release it tomorrow. 

Thanks for the suggestions.
